### PR TITLE
feat: [cheqd] add `top_accounts` module

### PR DESCRIPTION
**Pull request by huichiaotsou**
## Description

Closes: #XXXX

jira: https://forbole.atlassian.net/browse/BDU-707

can test on cosmoshub where there should be enough transactions

```yaml
chain:
    bech32_prefix: cosmos
    modules:
        - messages
        - auth
        - top_accounts
```
```yaml
node:
    type: remote 
    config:
        rpc:
            client_name: juno
            address: https://rpc.cosmoshub.forbole.com:443 
            max_connections: 25
        grpc:
            address: https://grpc.cosmoshub.forbole.com:443
            insecure: false
```
---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch
- [x] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: database/types/top_accounts.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		Reward:       reward,
+		Sum:          sum,
+	}
+}
+
+// Equals return true if one TopAccountsRow representing the same row as the original one
+func (a TopAccountsRow) Equals(b TopAccountsRow) bool {
+	return a.Address == b.Address &&
+		a.Available == b.Available &&
+		a.Redelegation == b.Redelegation &&
```
Comment (MonikaCat) [inline database/types/top_accounts.go:32]:
```suggestion
		a.Delegation == b.Delegation &&
		a.Redelegation == b.Redelegation &&
```

File: database/types/top_accounts.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+		a.Unbonding == b.Unbonding &&
+		a.Reward == b.Reward
```
Comment (MonikaCat) [inline database/types/top_accounts.go]:
```suggestion
		a.Reward == b.Reward &&
		a.Sum == b.Sum
```

File: modules/staking/utils_delegations.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"github.com/forbole/bdjuno/v3/modules/pricefeed"
+	"github.com/forbole/bdjuno/v3/types"
+	"github.com/rs/zerolog/log"
+)
+
+func (m *Module) RefreshDelegations(height int64, delegatorAddr string) error {
+	log.Debug().
+		Str("module", "staking").
+		Int64("height", height).Msg("updating delegation")
```
Comment (MonikaCat) [inline modules/staking/utils_delegations.go]:
```suggestion
		Int64("height", height).Msg("updating delegations")
```

File: modules/staking/utils_delegations.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+			coin = coin.Add(r.Balance)
+		}
+	}
+
+	err := m.db.SaveTopAccountsBalance("delegation",
+		[]types.NativeTokenAmount{
+			types.NewNativeTokenAmount(delegatorAddr, coin.Amount, height),
+		})
+	if err != nil {
+		return fmt.Errorf("error while savting top accounts delegation from MsgDelegate: %s", err)
```
Comment (MonikaCat) [inline modules/staking/utils_delegations.go]:
```suggestion
		return fmt.Errorf("error while saving top accounts delegation from MsgDelegate: %s", err)
```

File: modules/staking/utils_redelegations.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+			}
+		}
+	}
+
+	err := m.db.SaveTopAccountsBalance("redelegation",
+		[]types.NativeTokenAmount{
+			types.NewNativeTokenAmount(delegatorAddr, coin, height),
+		})
+	if err != nil {
+		return fmt.Errorf("error while savting top accounts delegation from MsgDelegate: %s", err)
```
Comment (MonikaCat) [inline modules/staking/utils_redelegations.go]:
```suggestion
		return fmt.Errorf("error while saving top accounts redelegation from MsgBeginRedelegate: %s", err)
```

File: modules/staking/utils_unbondings.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		res, err := m.source.GetUnbondingDelegations(
+			height,
+			delegatorAddr,
+			&query.PageRequest{
+				Key:   nextKey,
+				Limit: 100,
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("error while getting delegations: %s", err)
```
Comment (MonikaCat) [inline modules/staking/utils_unbondings.go:30]:
```suggestion
			return fmt.Errorf("error while getting unbonding delegations: %s", err)
```

File: modules/staking/utils_unbondings.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+			}
+		}
+	}
+
+	err := m.db.SaveTopAccountsBalance("unbonding",
+		[]types.NativeTokenAmount{
+			types.NewNativeTokenAmount(delegatorAddr, coin, height),
+		})
+	if err != nil {
+		return fmt.Errorf("error while savting top accounts delegation from MsgDelegate: %s", err)
```
Comment (MonikaCat) [inline modules/staking/utils_unbondings.go:48]:
```suggestion
		return fmt.Errorf("error while saving top accounts unbonding delegation from MsgUndelegate: %s", err)
```

File: modules/top_accounts/handle_periodic_operations.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+// RegisterPeriodicOperations implements modules.Module
+func (m *Module) RegisterPeriodicOperations(scheduler *gocron.Scheduler) error {
+	log.Debug().Str("module", "top accounts").Msg("setting up periodic tasks")
+
+	if _, err := scheduler.Every(1).Day().At("00:00").Do(func() {
+		utils.WatchMethod(m.RefreshRewards)
+	}); err != nil {
+		return fmt.Errorf("error while setting up bank periodic operation: %s", err)
```
Comment (MonikaCat) [inline modules/top_accounts/handle_periodic_operations.go]:
```suggestion
		return fmt.Errorf("error while setting up top accounts periodic operation: %s", err)
```

File: modules/top_accounts/handle_periodic_operations.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+	// Refresh rewards
+	err = m.distrModule.RefreshDelegatorRewards(height, delegators)
+	if err != nil {
+		return fmt.Errorf("error while refreshing delegators rewards: %s", err)
+	}
+
+	err = m.refreshTopAccountsSum(delegators)
+	if err != nil {
+		return fmt.Errorf("error while refreshing top accounts sum while handling MsgDelegate: %s", err)
```
Comment (MonikaCat) [inline modules/top_accounts/handle_periodic_operations.go]:
```suggestion
		return fmt.Errorf("error while refreshing top accounts sum value: %s", err)
```

File: modules/top_accounts/module.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/forbole/bdjuno/v3/database"
+
+	"github.com/forbole/juno/v3/modules"
+	junomessages "github.com/forbole/juno/v3/modules/messages"
+)
+
+var (
+	_ modules.Module        = &Module{}
+	_ modules.MessageModule = &Module{}
+	// _ modules.PeriodicOperationsModule = &Module{}
```
Comment (MonikaCat) [inline modules/top_accounts/module.go:15]:
```suggestion
	_ modules.PeriodicOperationsModule = &Module{}
```

File: modules/top_accounts/module.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+// Module represent x/gov module
```
Comment (MonikaCat) [inline modules/top_accounts/module.go:18]:
```suggestion
// Module represent top_accounts module
```

File: modules/top_accounts/utils_refresh.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+		}
+	}
+	return nil
+}
+
+func (m *Module) refreshDelegations(height int64, delegator string) func() {
+	return func() {
+		err := m.stakingModule.RefreshDelegations(height, delegator)
+		if err != nil {
+			log.Error().Str("module", "top acconts").Err(err).
```
Comment (MonikaCat) [inline modules/top_accounts/utils_refresh.go]:
```suggestion
			log.Error().Str("module", "top_accounts").Err(err).
```

File: modules/top_accounts/utils_refresh.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+				Str("operation", "refresh delegations").Msg("error while refreshing delegations")
+		}
+	}
+}
+
+func (m *Module) refreshRedelegations(tx *juno.Tx, delegatorAddr string) func() {
+	return func() {
+		err := m.stakingModule.RefreshRedelegations(tx.Height, delegatorAddr)
+		if err != nil {
+			log.Error().Str("module", "top acconts").Err(err).
```
Comment (MonikaCat) [inline modules/top_accounts/utils_refresh.go]:
```suggestion
			log.Error().Str("module", "top_accounts").Err(err).
```

File: modules/top_accounts/utils_refresh.go [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+				Str("operation", "refresh delegations").Msg("error while refreshing delegations")
```
Comment (MonikaCat) [inline modules/top_accounts/utils_refresh.go]:
```suggestion
				Str("operation", "refresh redelegations").Msg("error while refreshing redelegations")
```

File: modules/top_accounts/utils_refresh.go [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+		}
+	}
+}
+
+func (m *Module) refreshUnbondings(height int64, delegatorAddr string) func() {
+	return func() {
+		err := m.stakingModule.RefreshUnbondings(height, delegatorAddr)
+		if err != nil {
+			log.Error().Str("module", "top acconts").Err(err).
```
Comment (MonikaCat) [inline modules/top_accounts/utils_refresh.go]:
```suggestion
			log.Error().Str("module", "top_accounts").Err(err).
```

File: modules/top_accounts/utils_refresh.go [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+				Str("operation", "refresh unbondings").Msg("error while refreshing unbonding delegations")
+		}
+	}
+}
+
+func (m *Module) refreshBalance(height int64, address string) func() {
+	return func() {
+		err := m.bankModule.UpdateBalances([]string{address}, height)
+		if err != nil {
+			log.Error().Str("module", "top acconts").Err(err).
```
Comment (MonikaCat) [inline modules/top_accounts/utils_refresh.go]:
```suggestion
			log.Error().Str("module", "top_accounts").Err(err).
```

File: modules/top_accounts/utils_refresh.go [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+				Str("operation", "update balance").Msg("error while updating account available balances")
+		}
+
+		err = m.refreshTopAccountsSum([]string{address})
+		if err != nil {
+			log.Error().Str("module", "top acconts").Err(err).
```
Comment (MonikaCat) [inline modules/top_accounts/utils_refresh.go]:
```suggestion
			log.Error().Str("module", "top_accounts").Err(err).
```

File: modules/top_accounts/utils_refresh.go [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+				Str("operation", "update balance").Msg("error while refreshing top accounts sum while refreshing balance")
```
Comment (MonikaCat) [inline modules/top_accounts/utils_refresh.go]:
```suggestion
				Str("operation", "update top accounts sum").Msg("error while refreshing top accounts sum while refreshing balance")
```

File: types/bank.go [hunk 19]
```diff
@@ -17,3 +17,19 @@ func NewAccountBalance(address string, balance sdk.Coins, height int64) AccountB
...
 }
+
+// NativeTokenAmount represents the native token balance of an account at a given height
+type NativeTokenAmount struct {
+	Address string
+	Balance sdk.Int
+	Height  int64
+}
+
+// NewNativeTokenAmount allows to build a new NativeTokenBalance instance
```
Comment (MonikaCat) [inline types/bank.go]:
```suggestion
// NewNativeTokenAmount allows to build a new NativeTokenAmount instance
```

File: CHANGELOG.md [hunk 20]
```diff
@@ -1,6 +1,10 @@
 ## Unreleased
 ### Changes
 
+#### Top Accounts Module
+- ([\#510](https://github.com/forbole/bdjuno/pull/510)) Implemented top accounts module to store chain native token's balance for ranking
```
Comment (MonikaCat) [inline CHANGELOG.md]:
```suggestion
- ([\#510](https://github.com/forbole/bdjuno/pull/510)) Implemented `top_accounts` module to store chain native token's balance for ranking 
```

File: CHANGELOG.md [hunk 21]
```diff
@@ -1,6 +1,10 @@
 ## Unreleased
 ### Changes
 
+#### Top Accounts Module
```
Comment (MonikaCat) [inline CHANGELOG.md]:
```suggestion
#### [cheqd] Top Accounts Module
```

File: modules/top_accounts/handle_msg.go [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+
+	}
+
+	return nil
+}
+
+func (m *Module) handleMsgDelegate(height int64, delAddr string) error {
+	err := m.stakingModule.RefreshDelegations(height, delAddr)
+	if err != nil {
+		return fmt.Errorf("error while refreshing delegations: %s", err)
```
Comment (MonikaCat) [inline modules/top_accounts/handle_msg.go]:
```suggestion
		return fmt.Errorf("error while refreshing delegations while handling MsgDelegate: %s", err)
```

File: modules/top_accounts/handle_msg.go [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+	tx *juno.Tx, index int, delAddr string) error {
+
+	err := m.stakingModule.RefreshRedelegations(tx.Height, delAddr)
+	if err != nil {
+		return fmt.Errorf("error while refreshing redelegations while handling MsgBeginRedelegate: %s", err)
+	}
+
+	err = m.refreshTopAccountsSum([]string{delAddr})
+	if err != nil {
+		return fmt.Errorf("error while refreshing top accounts sum while handling MsgDelegate: %s", err)
```
Comment (MonikaCat) [inline modules/top_accounts/handle_msg.go]:
```suggestion
		return fmt.Errorf("error while refreshing top accounts sum while handling MsgBeginRedelegate: %s", err)
```

File: modules/top_accounts/handle_msg.go [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	err = m.bankModule.UpdateBalances([]string{delAddr}, height)
+	if err != nil {
+		return fmt.Errorf("error while updating account available balances with MsgWithdrawDelegatorReward: %s", err)
+	}
+
+	err = m.refreshTopAccountsSum([]string{delAddr})
+	if err != nil {
+		return fmt.Errorf("error while refreshing top accounts sum while handling MsgDelegate: %s", err)
```
Comment (MonikaCat) [inline modules/top_accounts/handle_msg.go]:
```suggestion
		return fmt.Errorf("error while refreshing top accounts sum while handling MsgWithdrawDelegatorReward: %s", err)
```

**Review by MonikaCat — APPROVED** [review]
LGTM

**Review by rmontagnin — APPROVED** [review]
utACK

**Pull request merged**